### ISSUE TITLE
Fetchgroup issue with derived metrics

### ISSUE THREAD

**Issue by myllynen** [performancecopilot/pcp#1253]
With the attached C reproducer only the last instance has a value reported for it when requesting derived metrics. If not using derived metrics then values are reported for all instances.

```
# gcc -Wall -lpcp pmfg-derived.c
# ./a.out 1 2 3                                                         
1:	-nan
2:	-nan
3:	476.880000
```
Thanks.

[pmfg-derived.c.txt](https://github.com/performancecopilot/pcp/files/6419555/pmfg-derived.c.txt)

**Comment by natoscott** [comment]
Try this patch @myllynen - seems to be some strange special-case handling in fetchgroup.c relating to derived metrics that AFAICS should not be there...

```
diff --git a/src/libpcp/src/fetchgroup.c b/src/libpcp/src/fetchgroup.c
index ddf4d64e9..3571387d7 100644
--- a/src/libpcp/src/fetchgroup.c
+++ b/src/libpcp/src/fetchgroup.c
@@ -192,18 +192,12 @@ pmfg_lookup_item(const char *metric, const char *instance, pmFGI item)
 	(item->u.item.metric_desc.indom != PM_INDOM_NULL && !instance))
 	return PM_ERR_INDOM;
 
-    /* Clear instance profile if this is an opaque derived metric. */
-    if (IS_DERIVED(item->u.item.metric_desc.pmid)) {
-        (void) pmAddProfile(PM_INDOM_NULL, 0, NULL);
-    }
-
     /* Add given instance to profile. */
     if (item->u.item.metric_desc.indom != PM_INDOM_NULL) {
 	sts = pmLookupInDom(item->u.item.metric_desc.indom, instance);
 	if (sts < 0)
 	    return sts;
 	item->u.item.metric_inst = sts;
-        /* Moot & harmless if IS_DERIVED. */
 	sts = pmAddProfile(item->u.item.metric_desc.indom, 1,
 			   &item->u.item.metric_inst);
     }
@@ -226,11 +220,6 @@ pmfg_lookup_indom(const char *metric, pmFGI item)
     if (sts < 0)
 	return sts;
 
-    /* Clear instance profile if this is an opaque derived metric. */
-    if (IS_DERIVED(item->u.item.metric_desc.pmid)) {
-        (void) pmAddProfile(PM_INDOM_NULL, 0, NULL);
-    }
-
     /* As a convenience to users, we also accept non-indom'd metrics */
     if (item->u.indom.metric_desc.indom == PM_INDOM_NULL)
 	return 0;
@@ -238,7 +227,6 @@ pmfg_lookup_indom(const char *metric, pmFGI item)
     /*
      * Add all instances; this will override any other past or future
      * piecemeal instance requests from __pmExtendFetchGroup_lookup.
-     * Moot & harmless if IS_DERIVED.
      */
     return pmAddProfile(item->u.indom.metric_desc.indom, 0, NULL);
 }
```

**Comment by myllynen** [comment]
This unfortunately doesn't seem to be this simple.

The above patches fixed the test program, yes.

But with the patch there is a tiny failure as part of QA 1069 added as part of commit 28f3a7f6 and more failures as part of QA 1140. There is a reference to issue 451 which has a link to mailing list post which discussed some related aspects.

Thanks.

**Comment by natoscott** [comment]
Ah, I see.  Sounds like the most reliable solution will be to remove the use of profiles entirely from the fetchgroup code.

**Comment by kmcdonell** [comment]
Fetchgroup really needs profiles to avoid unnecessary instances being fetched.

What fetchgroup does not need is any knowledge of derived metrics ... the code there at the moment is (a) unnecessary, and (b) wrong by inspection (it shotgun blasts the profile in several wrong ways).  I am assuming here that fetchgroup operations only make sense once you have a PMAPI context, by which time derived metrics are "bound" and indistinguishable to real metrics except deep in libpcp where the derived metric run-time magic happens.

If we remove all the references to derived metrics in the fetchgroup, and there is QA fallout then I think we need to look at those cases one-by-one as they are probably some other bug that is being masked by the fetchgroup's misuse of fetch profiles.

**Comment by kmcdonell** [comment]
As I suspected, there was another misuse of fetch profiles in fetchgroup.c, this time blasting all profiles under the sun in with this little gem in pmfg_clear_profile()
```return pmDelProfile(PM_INDOM_NULL, 0, NULL);```
replacing that with a sniper instead of a thermonuclear weapon, leaves us with:

1. check -g pmrep passing is passing all tests, including 1140 and 1069 that were victims of the blast above when the derived metric mumbo jumbo was removed form fetchgroup.c
2. Marko's test case passes (I'll make this a new QA test)
3. Assorted hand checking with pmrep using real metrics and derived metrics identical to the real metrics shows the same results.

Commits will be forthcoming.
